influence_benchmark/RL/expert_iteration.py

```diff
@@ -0,0 +1,211 @@
+import json
+import multiprocessing as mp
+import subprocess
+from collections import defaultdict
+from datetime import datetime
+from pathlib import Path
+
+import yaml
+
+from influence_benchmark.agent.hf_agent import HFAgent
+from influence_benchmark.backend.hf_backend import HFBackend
+from influence_benchmark.root import PROJECT_ROOT
+from influence_benchmark.vectorized_environment.vectorized_environment import VecEnv
+
+
+class ExpertIteration:
+    def __init__(
+        self,
+        env_args: dict,
+        training_args: dict,
+        accelerate_config: str,
+        sft_script_path: str,
+        model_name: str,
+        num_gen_trajectories: int,
+        num_chosen_trajectories: int,
+        iterations: int,
+        run_name: str = None,
+    ):
+        with open(accelerate_config, "r", encoding="utf-8") as f:
+            accelerate = yaml.safe_load(f)
+            self.devices = ["cuda:" + str(id) for id in accelerate["gpu_ids"] if id != ","]
+            print(self.devices)
+
+        assert num_gen_trajectories > (env_args["num_envs_per_device"] + 1) * len(
+            self.devices
+        ), "num_gen_trajectories must be higher than (num_envs_per_device +1) * num_devices"
+
+        if run_name is None:
+            self.run_name = datetime.now().strftime("%Y-%m-%d_%H-%M-%S")
+        else:
+            self.run_name = run_name
+        self.env_args = env_args
+        self.training_args = training_args
+        self.accelerate_config = accelerate_config
+        self.sft_script_path = sft_script_path
+
+        self.num_gen_trajectories = num_gen_trajectories
+        self.num_chosen_trajectories = num_chosen_trajectories
+        self.iterations = iterations
+
+        self.model_name = model_name
+        self.iteration_step = 0
+
+    def create_environment_and_agent(self, device, lora_path=None):
+        backend = HFBackend(self.model_name, device, lora_path=lora_path)
+        agent = HFAgent(self.env_args["env_name"], backend)
+        env_configs = [self.env_args] * self.env_args["num_envs_per_device"]
+        vec_env = VecEnv(env_configs=env_configs, backend=backend)
+        return vec_env, agent, backend
+
+    def create_backends_inference(self, lora_path=None):
+        self.backends = []
+        for device in self.devices:
+            self.backends.append(HFBackend(self.model_name, device, lora_path=lora_path))  # TODO fix
+
+    def create_agents_inference(self):
+        self.agents = []
+        for backend in self.backends:
+            self.agents.append(HFAgent(self.env_args["env_name"], backend))
+
+    def launch(self):
+        lora_path = None
+        for i in range(self.iterations):
+            trajectory_folder = Path(PROJECT_ROOT) / ".." / "data" / self.run_name / str(self.iteration_step)
+            trajectory_folder.mkdir(parents=True, exist_ok=True)
```

Comment:
Do we actually want `exist_ok`? Generally i think it's bad practice to overwrite previous results (can make it happen accidentally)